scripts/profiler/data_collector.py

```diff
@@ -47,47 +50,57 @@ def main():
     else:
         log_lvl_number = logging.INFO
 
-    # event is made to ensure that ModelCreator class is initialized before Rtt2Stream starts sending data
+    # Event is made to ensure that ModelCreator class is initialized before Rtt2Stream starts sending data
     event = Event()
+    # Setting these events results in closing corresponding modules.
+    event_close_rtt2stream = Event()
+    event_close_model_creator = Event()
 
     streams = Stream.create_stream(2)
 
-    try:
-        processes = []
-        processes.append(Process(target=rtt2stream,
-                                 args=(streams[0], event, log_lvl_number),
-                                 daemon=True))
-        processes.append(Process(target=model_creator,
-                                 args=(streams[1], event, args.dataset_name, log_lvl_number),
-                                 daemon=True))
-
-        for p in processes:
-            p.start()
-
-        start_time = time.time()
-        is_waiting = True
-        while is_waiting:
-            for p in processes:
-                p.join(timeout=0.5)
-                # Terminate other processes if one of the processes is not active.
-                if len(processes) > len(active_children()):
-                    is_waiting = False
-                    break
-
-                # Terminate every process on timeout.
-                if (time.time() - start_time) >= args.time:
-                    is_waiting = False
-                    break
-
-        for p in processes:
+    processes = []
+    processes.append((Process(target=rtt2stream,
+                                args=(streams[0], event, event_close_rtt2stream, log_lvl_number),
+                                daemon=True),
+                        event_close_rtt2stream))
+    processes.append((Process(target=model_creator,
+                                args=(streams[1], event, event_close_model_creator,
+                                    args.dataset_name, log_lvl_number),
+                                daemon=True),
+                        event_close_model_creator))
+
+    def close_processes():
+        for p, event_close in processes:
             if p.is_alive():
-                os.kill(p.pid, signal.SIGINT)
+                event_close.set()
                 # Ensure that we stop processes in order to prevent profiler data drop.
                 p.join()
 
-    except KeyboardInterrupt:
-        for p in processes:
-            p.join()
+    def signal_handler(sig, frame):
+        close_processes()
+        sys.exit()
+
+    signal.signal(signal.SIGINT, signal_handler)
```

Comment:
I would register the handler as early as possible (at the beginning of main). That could allow us to handle the SIGINT also when it arrives earlier.